browser/net/brave_ads_status_header_network_delegate_helper.cc

```diff
@@ -0,0 +1,46 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/net/brave_ads_status_header_network_delegate_helper.h"
+
+#include <string>
+#include <vector>
+
+#include "brave/browser/brave_ads/ads_service_factory.h"
+#include "brave/components/brave_ads/browser/ads_service.h"
+#include "brave/components/brave_search/common/brave_search_utils.h"
+#include "chrome/browser/profiles/profile.h"
+#include "content/public/browser/browser_context.h"
+#include "net/base/net_errors.h"
+
+namespace brave {
+
+namespace {
+
+constexpr char kAdsStatusHeader[] = "X-Brave-Ads-Enabled";
+constexpr char kAdsEnabledStatusValue[] = "1";
+
+}  // namespace
+
+int OnBeforeStartTransaction_AdsStatusHeader(
+    net::HttpRequestHeaders* headers,
+    const ResponseCallback& next_callback,
+    std::shared_ptr<BraveRequestInfo> ctx) {
+  brave_ads::AdsService* ads_service =
+      brave_ads::AdsServiceFactory::GetForProfile(
+          Profile::FromBrowserContext(ctx->browser_context));
+
+  if (!ads_service || !ads_service->IsEnabled() ||
+      !brave_search::IsAllowedHost(ctx->request_url)) {
+    return net::OK;
+  }
+
```

Comment:
should this first check if the main frame URL is also an allowed host before injecting the header?

for instance `search.brave.com/img.jpg` loaded on `example.com`  probably doesn't need the header? it seemed in all of your test cases the main frame URL is search.

Reply:
@remusao, do we have allowed list for a request origin hosts? Also could we not send header for sub-resources?

Reply:
@diracdeltas @aseren I believe we can restrict sending the headers to either top-level loads (e.g. visiting `https://search.brave.com/` or `https://search.brave.com/search?q=brave`) and for sub-resources originating from one of the Brave Search domains (e.g. `https://search.brave.com/img.jpg` would have the header if originating from `search.brave.com` but not from `example.com`).

The list of allowed top-level origins we'd like to allow could be `search.brave.com`, `search.brave.software`, `search.bravesoftware.com`, `safesearch.brave.com`, `safesearch.brave.software`, `safesearch.bravesoftware.com` and `search-dev-local.brave.com` (for our local setup). Cc @y3ti, did I forget anything?

I think sub-resources should have the security header preventing their inclusion on other origins anyway so we're probably not interested in these cases.

Reply:
Added check for tab origin:

```
  if (!ads_service || !ads_service->IsEnabled() ||
      !brave_search::IsAllowedHost(ctx->tab_origin) ||
      !brave_search::IsAllowedHost(ctx->request_url)) {
    return net::OK;
  }
```

Also added `safesearch.brave.com`, `safesearch.brave.software`, `safesearch.bravesoftware.com` to supported hosts list.

Reply:
> Cc @y3ti, did I forget anything?

I think you mentioned all domains :)